CMakeLists.txt

```diff
@@ -36,9 +35,9 @@ include_directories(include
   ${tf2_ros_INCLUDE_DIRS}
   ${laser_geometry_INCLUDE_DIRS}
   ${angles_INCLUDE_DIRS}
-  ${xmlrpcpp_INCLUDE_DIRS}
   ${rostime_INCLUDE_DIRS}
   ${TinyXML_INCLUDE_DIRS}
+  ${EIGEN3_INCLUDE_DIR}
```

Comment:
@bfjelds actually I think that even this might not be necessary; if arment is similar to catkin, then I believe laser_geometry should export the eigen dependency such that the eigen include directories end up in laser_geometry_INCLUDE_DIRS. @dirk-thomas if laser_geometry uses eigen headers, how should it export that? Does the export_dependencies do the trick? See https://github.com/ros-perception/laser_geometry/pull/27/files#diff-af3b638bc2a3e6c650974192a53c7291R45

Reply:
That makes sense ... makes me think that class_loader is probably not required either (shouldn't pluginlib export it as well?).

Reply:
> if laser_geometry uses eigen headers, how should it export that? Does the export_dependencies do the trick?

I don't think so. If the CMake variables don't follow the "convention" - you `find_package(Eigen3)` but the variables are named `EIGEN3_` - notice the difference case - I would expect that the include directories of `laser_geometry` won't include Eigen. You might need a CMake extra file to find_package Eigen and then add its include directories to the packages include dirs.

Reply:
Would it be better for laser_geometry to export the include dir like this:

```
ament_export_include_directories(include ${EIGEN3_INCLUDE_DIR})
```

Or for laser_filters to find_package(Eigen3)?

Reply:
> Would it be better for laser_geometry to export the include dir like this:

That makes the generated CMake config file contain the absolute path. In general it is preferred to resolve the dependency "on-demand" instead.

> Or for laser_filters to find_package(Eigen3)?

Downstream packages shouldn't need to find packages which are only transitive dependencies.

Reply:
OK.  I'm not too sure what "on-demand" means in this context, but I also tried adding something like this to laser_geometry's CMakeLists.txt:

```
# Eigen3 uses non-standard variable for
# include dirs (case and name): EIGEN3_INCLUDE_DIR.
if(NOT Eigen3_INCLUDE_DIRS)
  if (EIGEN3_INCLUDE_DIR)
    message("use EIGEN3_INCLUDE_DIR (${EIGEN3_INCLUDE_DIR}) for Eigen3_INCLUDE_DIRS")
    set(Eigen3_INCLUDE_DIRS ${EIGEN3_INCLUDE_DIR})
    message("Eigen3_INCLUDE_DIRS=${Eigen3_INCLUDE_DIRS}")
  else()
    message("Eigen3_INCLUDE_DIRS not found")
  endif()
endif()
```

In looking at how ament's export works, it seems like this "translation" should work (and anecdotally, the build works with it).

Reply:
@bfjelds sorry I haven't had time to look deeper into this. Here's my undesrtanding so far:

- On kinetic (and other recent rosdistros, laser_geometry has used `FindPackage(eigen)` which has resolved to the ROS-specific cmake file for eigen here: https://github.com/ros/cmake_modules/blob/0.4-devel/cmake/Modules/FindEigen.cmake
- That isn't actually best practice. Since ROS Jade, packages are encouraged to directly `FindPackage(Eigen3)` as described here: http://wiki.ros.org/jade/Migration#Eigen_CMake_Module_in_cmake_modules . This is complicated because Eigen uses non standard variable names. The recommended workaround in the previous link for catkin is to explicitly add eigen's include directories to the INCLUDE_DIRS exported by the package using it.
- I think that @dirk-thomas is suggesting adding a CMake extra file to laser_geometry. That file would then be run whenever somebody does `FindPackage(laser_geometry)`, and it would call `FindPackage(Eigen3)` and add those include directories to laser_geometry_INCLUDE_DIRS.

@dirk-thomas does that summary sound right? @bfjelds I wasn't able to find any ament-specific documentation on adding a CMakeExtras file, but I think you should be able to create one for laser_geometry by following the example rcl_cpp-extras.cmake in the rcl_cpp package: https://github.com/ros2/rclcpp/tree/b81f55e5df0c3a43ba454b741e772ba77de7375b/rclcpp

Thanks for working on this @bfjelds . You're blazing a good path - working through lots of bugs since this is new.

Reply:
> does that summary sound right?

:+1: